# Assignments: Loops and Control Structures

**Name:** T Lakshman Aditya

**Date:** 12-2-2026

---

## Instructions
- Complete all the problems below
- Write your code in the provided cells
- Test your solutions with different inputs
- Comment your code where necessary
- Each problem has a difficulty level: Easy, Medium, or Hard

---

## Section 1: Control Structures (If-Else Statements)

### Problem 1.1: Grade Calculator (Easy)

Write a program that takes a student's marks (0-100) as input and prints their grade according to the following criteria:
- 90-100: Grade A
- 80-89: Grade B
- 70-79: Grade C
- 60-69: Grade D
- Below 60: Grade F

Also, validate that the marks are within the valid range (0-100).

**Hint:** Use if-elif-else statements. First check if marks are in valid range (0-100), then use >= comparisons starting from highest grade.

In [3]:
marks = int(input("Enter Marks"))

if marks>=90:
    print("Grade A")
elif marks>=80:
    print("Grade B")
elif marks>=70:
    print("Grade C")
elif marks>=60:
    print("Grade D")
elif marks<60:
    print("Grade F") 

Grade B


### Problem 1.2: Leap Year Checker (Medium)

Write a program to check if a given year is a leap year or not.

Rules:
- A year is a leap year if it is divisible by 4
- However, if it is divisible by 100, it is NOT a leap year
- Exception: If it is also divisible by 400, then it IS a leap year

Examples:
- 2000: Leap year (divisible by 400)
- 1900: Not a leap year (divisible by 100 but not 400)
- 2024: Leap year (divisible by 4 but not 100)

**Hint:** Check conditions in order: first divisible by 400, then by 100, then by 4. Use modulo operator (%) to check divisibility.

In [4]:
year = int(input("Enter Year"))
if (year % 4 == 0 and year % 100 != 0)  or (year % 400 == 0):
    print(f"your year {year} is leap year ")
else:
    print(f"your year {year} is not leap year ")


your year 2000 is leap year 


### Problem 1.3: Triangle Type Identifier (Medium)

Write a program that takes three sides of a triangle as input and determines:
1. Whether the three sides can form a valid triangle (sum of any two sides must be greater than the third side)
2. If valid, classify the triangle as:
   - Equilateral (all three sides equal)
   - Isosceles (exactly two sides equal)
   - Scalene (no sides equal)

**Hint:** First check: (a+b>c) and (b+c>a) and (a+c>b). Then use == to compare sides: if all three equal, it's equilateral; if any two equal, it's isosceles; otherwise scalene.

In [12]:
a = int(input("Enter First side: "))
b = int(input("Enter Second side: "))
c = int(input("Enter Third side: "))
if (a + b > c) and (a + c > b) and (b + c > a):
    if a == b == c:
        print("It is an Equilateral Triangle")
    elif a == b or b == c or a == c:
        print("It is an Isosceles Triangle")
    else:
        print("It is a Scalene Triangle")
else:
    print("Not a valid Triangle")

print(f"a is {a}, b is {b} , c is {c}")


It is an Isosceles Triangle
a is 6, b is 5 , c is 6


### Problem 1.4: BMI Calculator with Categories (Easy)

Write a program that calculates BMI (Body Mass Index) and categorizes it.

Formula: BMI = weight (kg) / (height (m))²

Categories:
- Below 18.5: Underweight
- 18.5-24.9: Normal weight
- 25-29.9: Overweight
- 30 and above: Obese

**Hint:** First calculate BMI using the formula (use ** for power or multiply height by itself). Then use if-elif-else to categorize.

In [ ]:
# height in meters
weight = int(input("Enter Weight :"))
height = float(input("Enter height in meters"))

BMI = weight/(height * height)
print(f"BMI is {BMI:.2f}")

print(f"your weight is {weight}")
print(f"Your height is {height}")
if BMI < 18.5:
    print("Under Weight")
elif BMI < 24.9:
    print("Normal Weight")
elif BMI < 29.9:
    print("Over Weight")
elif BMI > 30:
    print("obese")


BMI is 28.49
Over Weight


In [ ]:
# code when height is in feet and inches
weight = float(input("Enter weight (kg): "))
feet = int(input("Enter height (feet): "))
inches = int(input("Enter remaining inches: "))

# Convert height to meters
total_inches = (feet * 12) + inches
height_meters = total_inches * 0.0254   # 1 inch = 0.0254 meters

print(f"your weight is {weight}")
print(f"Your height is {height_meters}")

# Calculate BMI
bmi = weight / (height_meters ** 2)

print(f"\nYour BMI is: {bmi:.2f}")

# Categorize
if bmi < 18.5:
    print("Category: Underweight")
elif bmi < 25:
    print("Category: Normal Weight")
elif bmi < 30:
    print("Category: Overweight")
else:
    print("Category: Obese")


your weight is 85.0
Your height is 1.7271999999999998

Your BMI is: 28.49
Category: Overweight


---

## Section 2: For Loops

### Problem 2.1: Multiplication Table (Easy)

Write a program that prints the multiplication table for a given number (1 to 10).

Example output for number 5:
```
5 x 1 = 5
5 x 2 = 10
5 x 3 = 15
...
5 x 10 = 50
```

**Hint:** Use for loop with range(1, 11) to iterate from 1 to 10. Multiply the number by loop variable i.

In [15]:
a = int(input("Enter number"))
for i in range(1,11):
    print(f"{a} x {i} = {a*i}")


5 x 1 = 5
5 x 2 = 10
5 x 3 = 15
5 x 4 = 20
5 x 5 = 25
5 x 6 = 30
5 x 7 = 35
5 x 8 = 40
5 x 9 = 45
5 x 10 = 50


### Problem 2.2: Factorial Calculator (Medium)

Write a program to calculate the factorial of a number using a for loop.

Factorial of n (n!) = n × (n-1) × (n-2) × ... × 1

Example: 5! = 5 × 4 × 3 × 2 × 1 = 120

**Hint:** Initialize result = 1. Use for loop with range(1, n+1) and multiply result by each number.

In [24]:
n = int(input("Enter number"))
result = 1
for i in range(1,n+1):
    result *= i
print(result)


120


### Problem 2.3: Sum of Even Numbers (Easy)

Write a program that calculates the sum of all even numbers between 1 and n (inclusive), where n is provided by the user.

Example: If n = 10, sum = 2 + 4 + 6 + 8 + 10 = 30

**Hint:** Use for loop with range(1, n+1). Check if i % 2 == 0, then add to sum. Or use range(2, n+1, 2) to iterate only even numbers.

In [7]:
numbers  = int(input())
for i in range(2,numbers+1,2):
    print(i)


2
4
6
8
10
12
14


In [9]:
numbers  = int(input())
for i in range(1,numbers+1):
    if i%2==0:
        print(i)

2
4
6
8
10
12
14


### Problem 2.4: Fibonacci Series (Medium)

Write a program to print the first n terms of the Fibonacci series.

The Fibonacci series: 0, 1, 1, 2, 3, 5, 8, 13, 21, ...

Each number is the sum of the two preceding ones.

**Hint:** Initialize a=0, b=1. Print a, then in loop: calculate next=a+b, print next, then update a=b and b=next.

In [11]:
n = int(input())
a = 0
b = 1
for i in range(n):
    print(a,end=" ")
    next = a+b
    a=b
    b=next


0 1 1 2 3 

In [14]:
n = int(input())
a,b =0,1
for i in range(n):
    print(a,end= " ")
    a,b=b,a+b


0 1 1 2 3 

### Problem 2.5: Prime Number Checker (Medium)

Write a program to check if a given number is prime or not.

A prime number is a natural number greater than 1 that has no positive divisors other than 1 and itself.

**Hint:** Check if n is divisible by any number from 2 to n-1. If no divisor found, it's prime. Optimize: check only up to √n.

In [ ]:
num1 = int(input())
count = 0
for i in range(2,num1):
    if num1 % i == 0:
        count +=1
if count == 0:
    print("yes number is prime")
else:
    print("no number is not prime")


yes


### Problem 2.6: Count Digits in a Number (Easy)

Write a program that counts the number of digits in a given integer.

Example: 12345 has 5 digits

**Hint:** Convert number to string and use len(), or keep dividing by 10 until number becomes 0, counting each division.

In [27]:
num1= int(input())
count = 0
if num1 == 0:
    count =1
while num1>0:
    count +=1
    num1 = num1 // 10

print(count)

2


---

## Section 3: While Loops

### Problem 3.1: Reverse a Number (Medium)

Write a program that reverses the digits of a given number using a while loop.

Example: Input: 12345, Output: 54321

**Hint:** Initialize reverse=0. While n>0: get last digit (n%10), add to reverse (reverse=reverse*10+digit), remove last digit (n=n//10).

In [8]:
num1 = int(input())
rev = 0
print(num1)
while num1 > 0:
    digit = num1 % 10
    rev = rev * 10 + digit
    num1 = num1 // 10
print(rev)

4569
9654


### Problem 3.2: Sum of Digits (Easy)

Write a program that calculates the sum of all digits in a given number.

Example: Input: 12345, Output: 15 (1+2+3+4+5)

**Hint:** While n>0: add last digit to sum (sum += n%10), then remove it (n = n//10).

In [9]:
num1 = int(input())
tot_sum = 0
print(num1)
while num1 > 0:
    digit = num1 % 10
    tot_sum = tot_sum + digit
    num1 = num1//10
print(tot_sum)

12345
15


### Problem 3.3: Number Guessing Game (Medium)

Write a number guessing game where:
1. The program generates a random number between 1 and 100
2. The user has to guess the number
3. After each guess, the program tells the user if their guess is too high or too low
4. The game continues until the user guesses correctly
5. Display the number of attempts taken

**Hint:** Use `import random` and `random.randint(1, 100)` to generate number. Use while loop with condition guess != target. Count attempts.

In [13]:
import random

target = random.randint(1,100)
guess = 0
attempts = 0
while guess != target:
    guess = int(input("Enter num from 1 to 100"))
    attempts += 1

    if guess < target:
        print("Too low")
    elif guess > target:
        print("Too high") 
    else:
        print("crt")
print(f"Total Num of attempts are {attempts}")

Too low
Too high
Too low
Too low
Too low
crt
Total Num of attempts are 6


### Problem 3.4: Palindrome Checker (Medium)

Write a program to check if a given number is a palindrome or not.

A palindrome number reads the same forwards and backwards.

Examples: 121, 12321, 1331 are palindromes

**Hint:** Store original number, reverse it (like Problem 3.1), then compare original with reversed number.

In [23]:
num1 = int(input("Enter num"))
num2 = num1
rev_num = 0
while num1>0:
    digit = num1%10
    rev_num = rev_num*10 + digit
    num1 = num1//10

if(num2 == rev_num):
    print(f"{num2} and {rev_num} are pallindrome")
else:
    print(f"{num2} and {rev_num} are not pallindrome")

1245 and 5421 are not pallindrome


### Problem 3.5: Armstrong Number (Hard)

Write a program to check if a number is an Armstrong number (also called Narcissistic number).

An Armstrong number is a number that is equal to the sum of its own digits each raised to the power of the number of digits.

Examples:
- 153 = 1³ + 5³ + 3³ = 1 + 125 + 27 = 153 (Armstrong number)
- 9474 = 9⁴ + 4⁴ + 7⁴ + 4⁴ = 6561 + 256 + 2401 + 256 = 9474 (Armstrong number)

**Hint:** First count digits (power). Store original. Extract each digit, raise to power, add to sum. Compare sum with original.

In [29]:
num1 = int(input("enter number"))
num2 = num1
temp = num1
tot_sum = 0
digit = 0
while num1 > 0:
    num1 = num1//10
    digit +=1
while num2 > 0:
    x = num2%10
    tot_sum += x**digit
    num2 = num2//10
if temp==tot_sum:
    print("They are Armstrong Number")
else:
    print("They are not Armstrong Number")


They are Armstrong Number


---

## Section 4: Nested Loops

### Problem 4.1: Pattern Printing - Right Triangle (Easy)

Print the following pattern for n rows:

```
*
**
***
****
*****
```
(Example shown for n=5)

**Hint:** Use nested loops: outer loop for rows (i from 1 to n), inner loop for columns (j from 1 to i). Print * in inner loop.

In [40]:
num1 = int(input())
for i in range(num1):
    for j in range(i+1):
        print("*",end= " ")
    print()


* 
* * 
* * * 
* * * * 
* * * * * 


### Problem 4.2: Pattern Printing - Inverted Right Triangle (Easy)

Print the following pattern for n rows:

```
*****
****
***
**
*
```
(Example shown for n=5)

**Hint:** Outer loop for rows. Inner loop prints stars from 1 to (n-i+1), where i is the row number.

In [45]:
num1 = int(input())
for i in range(num1,0,-1):
    for j in range(i):
        print("*",end= " ")
    print()

* * * * * 
* * * * 
* * * 
* * 
* 


### Problem 4.3: Pattern Printing - Pyramid (Medium)

Print the following pyramid pattern for n rows:

```
    *
   ***
  *****
 *******
*********
```
(Example shown for n=5)

**Hint:** For row i: print (n-i) spaces, then print (2*i-1) stars. Use nested loops or string multiplication.

In [55]:
n = int(input())

for i in range(1, n+1):
    for j in range(n-i):
        print(" ", end="")
        
    for k in range(2*i-1):
        print("*", end="")
        
    print()

    *
   ***
  *****
 *******
*********


### Problem 4.4: Pattern Printing - Number Pattern (Medium)

Print the following number pattern for n rows:

```
1
12
123
1234
12345
```
(Example shown for n=5)

**Hint:** Similar to Problem 4.1, but print j (column number) instead of *.

In [64]:
n = int(input())
for i in range(1,n+1):
    for j in range(i):
        print(j+1,end= " ")
    print()


1 
1 2 
1 2 3 
1 2 3 4 
1 2 3 4 5 


### Problem 4.5: Multiplication Table Grid (Medium)

Print a multiplication table grid from 1 to n.

Example for n=5:
```
1   2   3   4   5
2   4   6   8  10
3   6   9  12  15
4   8  12  16  20
5  10  15  20  25
```

**Hint:** Nested loops: for i in range(1, n+1) and for j in range(1, n+1). Print i*j with proper spacing.

In [70]:
num1 = int(input())
for i in range(1,n+1):
    for j in range(1,n+1):
        print(i*j,end= " ")
    print()


1 2 3 4 5 
2 4 6 8 10 
3 6 9 12 15 
4 8 12 16 20 
5 10 15 20 25 


### Problem 4.6: Diamond Pattern (Hard)

Print a diamond pattern for n rows (n should be odd):

```
    *
   ***
  *****
 *******
*********
 *******
  *****
   ***
    *
```
(Example shown for n=9)

**Hint:** Print pyramid (upper half) for rows 1 to (n+1)//2, then print inverted pyramid (lower half) for remaining rows.

In [76]:
n = 9
for i in range(1, n+1):
    for j in range(n-i):
        print(" ", end="")
    
    for k in range(1,(n+1)//2):
        print("*", end="")
        
    print()

        ****
       ****
      ****
     ****
    ****
   ****
  ****
 ****
****


---

## Section 5: Break and Continue Statements

### Problem 5.1: Find First Multiple (Easy)

Write a program that finds and prints the first number in the range 1 to 100 that is divisible by both 7 and 13. Use the `break` statement to exit the loop once found.

**Hint:** Loop through 1 to 100. Check if i%7==0 and i%13==0. If true, print i and use break to exit loop.

In [10]:
# Your code here
for i in range(1,100):
    if i%7==0 and i%13==0:
        print(i)
        break
else:
    print("no num found")

91


### Problem 5.2: Skip Multiples (Easy)

Write a program that prints all numbers from 1 to 50, but skips (doesn't print) numbers that are multiples of 3. Use the `continue` statement.

**Hint:** Loop from 1 to 50. If i%3==0, use continue to skip to next iteration. Otherwise print i.

In [1]:
# Your code here
for i in range(1,50):
    if i%3==0:
        pass
    else:
        print(i)

1
2
4
5
7
8
10
11
13
14
16
17
19
20
22
23
25
26
28
29
31
32
34
35
37
38
40
41
43
44
46
47
49


### Problem 5.3: Sum Until Negative (Medium)

Write a program that:
1. Keeps asking the user to enter numbers
2. Adds positive numbers to a sum
3. Skips zero (continues to next iteration)
4. Stops when a negative number is entered
5. Prints the final sum

**Hint:** Use while True loop. If num < 0, break. If num == 0, continue. Otherwise add to sum.

In [14]:
# Your code here
total = 0
while True:
    num1 = int(input("Enter the number"))

    if num1 < 0:
        break
    elif num1 == 0:
        continue
    else:
        total += num1

print(total) 
    


9


---

## Section 6: Combined Challenges

### Problem 6.1: Prime Numbers in Range (Medium)

Write a program to print all prime numbers between two given numbers (inclusive).

Example: Between 10 and 30: 11, 13, 17, 19, 23, 29

**Hint:** Outer loop for each number in range. For each number, use inner loop to check if it's prime (like Problem 2.5).

In [20]:
for i in range(10,31):
    for j in range (2,int(i**0.5)+1):
        if i%j == 0:
            break
    else:
        print(i)

11
13
17
19
23
29


### Problem 6.2: GCD Calculator (Medium)

Write a program to find the Greatest Common Divisor (GCD) of two numbers using the Euclidean algorithm.

Algorithm:
1. Divide the larger number by the smaller number
2. Replace the larger number with the smaller number
3. Replace the smaller number with the remainder
4. Repeat until the remainder is 0
5. The GCD is the last non-zero remainder

**Hint:** While b != 0: store remainder (a%b), then a=b, b=remainder. When loop ends, a is the GCD.

In [ ]:
# Your code here
a = int(input())
b = int(input())

while b!=0:
    remainder =  a%b
    a=b
    b = remainder
print(a)

6


### Problem 6.3: Perfect Number Checker (Hard)

Write a program to check if a number is a perfect number.

A perfect number is a positive integer that is equal to the sum of its proper positive divisors (excluding the number itself).

Example: 6 is perfect because 6 = 1 + 2 + 3

Example: 28 is perfect because 28 = 1 + 2 + 4 + 7 + 14

**Hint:** Loop from 1 to n-1. If i divides n (n%i==0), add i to sum. Compare sum with n.

In [31]:
# Your code here
a = int(input())
sum = 0
for i in range(1,a//2 +1): # n//2 it remove last digit if a is odd 
    if a%i==0:
        sum += i

if sum == a:
    print("perfect number")
else:
    print("not perfect")

perfect number


### Problem 6.4: LCM Calculator (Medium)

Write a program to find the Least Common Multiple (LCM) of two numbers.

Hint: LCM(a, b) = (a × b) / GCD(a, b)

You can use the GCD function from Problem 6.2.

**Hint:** First calculate GCD using Euclidean algorithm (Problem 6.2), then use formula: LCM = (a * b) // GCD.

In [38]:
a = int(input())
b = int(input())

x = a
y = b

while y != 0:
    remainder = x % y
    x = y
    y = remainder
gcd = x
lcm = (a * b) // gcd
print("LCM =", lcm)

LCM = 36


### Problem 6.5: Number to Words (Hard)

Write a program that converts a number (0-999) to words.

Examples:
- 0: Zero
- 25: Twenty Five
- 123: One Hundred Twenty Three
- 500: Five Hundred

**Hint:** Create lists for ones, teens, tens, etc. Extract hundreds, tens, ones digits. Convert each part to words and combine.

In [41]:
# Your code here
num = int(input("Enter a number (0-999): "))

ones = ["","One","Two","Three","Four","Five","Six","Seven","Eight","Nine"]
teens = ["Ten","Eleven","Twelve","Thirteen","Fourteen","Fifteen","Sixteen","Seventeen","Eighteen","Nineteen"]
tens = ["","","Twenty","Thirty","Forty","Fifty","Sixty","Seventy","Eighty","Ninety"]

if num == 0:
    print("Zero")
else:
    words = ""

    # Hundreds
    if num >= 100:
        words += ones[num//100] + " Hundred "
        num = num % 100

    # Teens
    if 10 <= num <= 19:
        words += teens[num-10]

    else:
        # Tens
        words += tens[num//10] + " "

        # Ones
        words += ones[num%10]

    print(words)

Four Hundred Fifty Six


### Problem 6.6: Strong Number Checker (Hard)

Write a program to check if a number is a Strong number.

A Strong number is a number whose sum of factorial of digits is equal to the original number.

Example: 145 is a strong number because 1! + 4! + 5! = 1 + 24 + 120 = 145

**Hint:** Extract each digit, calculate its factorial, add to sum. Compare sum with original number.

In [8]:
a = int(input())
temp = a 
total= 0
while temp >0:
    digit = temp%10

    fact = 1
    for i in range(1,digit+1):
        fact *= i
        
    total += fact
    temp //= 10

if total == a:
    print("yes")
else:
    print("No")


yes


### Problem 6.7: Pascal's Triangle (Hard)

Write a program to print Pascal's Triangle up to n rows.

Example for n=5:
```
    1
   1 1
  1 2 1
 1 3 3 1
1 4 6 4 1
```

Each number is the sum of the two numbers directly above it.

**Hint:** Use formula: C(n,k) = C(n,k-1) * (n-k+1) / k, where C(n,0)=1. Or use nested loops with list to store previous row.

In [56]:
n = int(input("Enter number of rows: "))

for i in range(n):
    num = 1
    
    # spaces for triangle shape
    print(" " * (n - i), end="")
    
    for j in range(i + 1):
        print(num, end=" ")
        
        num = num * (i - j) // (j + 1)
        
    print()

     1 
    1 1 
   1 2 1 
  1 3 3 1 
 1 4 6 4 1 


### Problem 6.8: Simple Calculator (Medium)

Create a simple calculator program that:
1. Shows a menu of operations: Add, Subtract, Multiply, Divide, Exit
2. Takes two numbers as input
3. Performs the selected operation
4. Shows the result
5. Repeats until the user chooses to exit
6. Handles division by zero error

**Hint:** Use while loop. Display menu, get choice. Use if-elif to perform operation. Check if divisor is 0 before division.

In [ ]:
while True:
    print("\n---Simple Calculator---")
    print("1.ADD")
    print("2.Subtract")
    print("3.Multiply")
    print("4.Divide")
    print("5.Exit")

    choice = int(input("Enter your Choice(1-5): "))
    if choice ==5:
        print("Extiing Calculator")
        break

    num1 = float(input("Enter 1st num"))
    num2 = float(input("Enter 2nd num"))
    
    if choice == 1:
        print("Result:",num1+num2)
    elif choice == 2:
        print("Result:",num1-num2)
    elif choice == 3:
        print("Result:",num1*num2)
    elif choice == 4:
        if num2 ==0:
            print("Error: Division by zero is not allowed")
        else:
            print("Result:",num1/num2)
    else:
        print("Invalid Choice")
    
    
    break


---Simple Calculator---
1.ADD
2.Subtract
3.Multiply
4.Divide
5.Exit
Result: 1.5


---

## Section 7: Bonus Challenges

### Problem 7.1: Collatz Conjecture (Medium)

The Collatz conjecture is one of the most famous unsolved problems in mathematics. The rules are:
1. Start with any positive integer n
2. If n is even, divide it by 2
3. If n is odd, multiply it by 3 and add 1
4. Repeat until you reach 1

Write a program that:
- Takes a number as input
- Shows the complete sequence until it reaches 1
- Counts how many steps it took

Example: Starting with 6: 6 → 3 → 10 → 5 → 16 → 8 → 4 → 2 → 1 (8 steps)

**Hint:** While n != 1: if n%2==0, n=n//2, else n=3*n+1. Count steps and print each value.

In [ ]:
# Your code here


### Problem 7.2: Caesar Cipher (Hard)

Write a program that implements a Caesar cipher to encrypt and decrypt messages.

The Caesar cipher shifts each letter by a fixed number of positions in the alphabet.

Example: With shift=3
- 'A' becomes 'D'
- 'Hello' becomes 'Khoor'

Your program should:
1. Ask for a message and shift value
2. Provide options to encrypt or decrypt
3. Handle both uppercase and lowercase letters
4. Keep non-alphabetic characters unchanged

**Hint:** Loop through each character. Use ord() to get ASCII value, shift it, use chr() to convert back. Use modulo 26 for wrapping. Check if char.isalpha().

In [ ]:
# Your code here


### Problem 7.3: ATM Machine Simulator (Hard)

Create an ATM machine simulator with the following features:
1. Initial balance: 10000
2. PIN: 1234 (3 attempts allowed)
3. Menu options:
   - Check Balance
   - Withdraw Money (check for sufficient balance and valid amount)
   - Deposit Money
   - Change PIN
   - Exit
4. Keep transaction history (last 5 transactions)
5. Input validation for all operations

**Hint:** Use variables for balance, pin. Loop for PIN attempts. Use while loop for menu. Use list to store transaction history. Validate amounts > 0 and balance >= withdrawal.

In [ ]:
# Your code here


---

## Section 8: Additional If-Else Practice (30 Questions)

### Problem 8.1: Positive, Negative, or Zero (Easy)

Write a program that checks if a number is positive, negative, or zero.

**Hint:** Use if-elif-else: if n > 0 (positive), elif n < 0 (negative), else (zero).

In [ ]:
# Your code here


### Problem 8.2: Even or Odd (Easy)

Write a program that determines if a number is even or odd.

**Hint:** If n % 2 == 0, it's even; otherwise odd.

In [ ]:
# Your code here


### Problem 8.3: Largest of Three Numbers (Easy)

Write a program to find the largest of three numbers.

**Hint:** Compare: if a >= b and a >= c, largest is a. Use nested if-else or use max() function.

In [ ]:
# Your code here


### Problem 8.4: Voting Eligibility (Easy)

Check if a person is eligible to vote (age >= 18).

**Hint:** If age >= 18, print eligible; else not eligible.

In [ ]:
# Your code here


### Problem 8.5: Character Type Checker (Medium)

Write a program that checks if a character is a vowel, consonant, digit, or special character.

**Hint:** Check if ch in 'aeiouAEIOU' (vowel), elif ch.isalpha() (consonant), elif ch.isdigit() (digit), else special.

In [ ]:
# Your code here


### Problem 8.6: Discount Calculator (Medium)

Calculate discount based on purchase amount:
- Below 1000: No discount
- 1000-5000: 10% discount
- 5000-10000: 20% discount
- Above 10000: 30% discount

**Hint:** Use if-elif-else with ranges. Calculate: final_amount = amount - (amount * discount_percentage / 100).

In [ ]:
# Your code here


### Problem 8.7: Quadrant Finder (Medium)

Given x and y coordinates, determine which quadrant the point lies in (or on an axis).

**Hint:** If x>0 and y>0: Quadrant I. If x<0 and y>0: Quadrant II. Similar for III and IV. Check for axes when x==0 or y==0.

In [ ]:
# Your code here


### Problem 8.8: Age Category (Easy)

Categorize age as: Child (0-12), Teenager (13-19), Adult (20-59), Senior (60+).

**Hint:** Use if-elif-else with age ranges. Check from Child to Senior in order.

In [ ]:
# Your code here


### Problem 8.9: Temperature Converter (Medium)

Convert temperature between Celsius and Fahrenheit based on user choice.
- C to F: F = (C × 9/5) + 32
- F to C: C = (F - 32) × 5/9

**Hint:** Ask user for choice (1 or 2). Use if-else to apply appropriate formula. Use parentheses correctly in calculations.

In [ ]:
# Your code here


### Problem 8.10: Password Strength Checker (Medium)

Check password strength:
- Weak: length < 6
- Medium: length 6-10
- Strong: length > 10

**Hint:** Use len(password) to get length. Compare with if-elif-else.

In [ ]:
# Your code here


### Problem 8.11: Divisibility Checker (Easy)

Check if a number is divisible by both 3 and 5, only 3, only 5, or neither.

**Hint:** Check if n%3==0 and n%5==0 (both), elif n%3==0 (only 3), elif n%5==0 (only 5), else (neither).

In [ ]:
# Your code here


### Problem 8.12: Movie Ticket Pricing (Medium)

Calculate ticket price:
- Children (< 12): $10
- Adults (12-59): $15
- Seniors (60+): $12
- Weekend: Add $3 to all prices

**Hint:** First determine base price by age. Then check if weekend and add $3 if true.

In [ ]:
# Your code here
 

### Problem 8.13: Letter Grade to GPA (Medium)

Convert letter grade to GPA: A=4.0, B=3.0, C=2.0, D=1.0, F=0.0

**Hint:** Use if-elif-else or dictionary: grade_dict = {'A': 4.0, 'B': 3.0, ...}. Then print grade_dict[letter].

In [ ]:
# Your code here


### Problem 8.14: Electricity Bill Calculator (Medium)

Calculate bill based on units consumed:
- 0-100: $0.50 per unit
- 101-200: $0.75 per unit
- 201-300: $1.20 per unit
- Above 300: $1.50 per unit

**Hint:** Use if-elif-else to determine rate based on units. Calculate: bill = units * rate.

In [ ]:
# Your code here


### Problem 8.15: Traffic Light System (Easy)

Based on light color, print action: Red=Stop, Yellow=Caution, Green=Go

**Hint:** Use if-elif-else or dictionary to map color to action.

In [ ]:
# Your code here


### Problem 8.16: Smallest of Three Numbers (Easy)

Find the smallest of three numbers.

**Hint:** Similar to Problem 8.3, but check for smallest. Or use min(a, b, c).

In [ ]:
# Your code here


### Problem 8.17: Number in Range (Easy)

Check if a number is between 1 and 100 (inclusive).

**Hint:** Check if 1 <= n <= 100.

In [ ]:
# Your code here


### Problem 8.18: Absolute Value (Easy)

Calculate the absolute value of a number without using built-in functions.

**Hint:** If n < 0, return -n; else return n.

In [ ]:
# Your code here


### Problem 8.19: Days in Month (Medium)

Given month number (1-12), print the number of days. Consider leap year for February.

**Hint:** Use if-elif for each month. For Feb (month==2), check if leap year (use Problem 1.2 logic) to return 29 or 28.

In [ ]:
# Your code here


### Problem 8.20: Profit or Loss Calculator (Easy)

Calculate profit or loss given cost price and selling price.

**Hint:** If selling_price > cost_price: profit = selling_price - cost_price. Else if less: loss. Else: no profit no loss.

In [ ]:
# Your code here


### Problem 8.21: Uppercase or Lowercase (Easy)

Check if a character is uppercase or lowercase.

**Hint:** Use ch.isupper() or ch.islower(), or compare ASCII values.

In [ ]:
# Your code here


### Problem 8.22: Roots of Quadratic Equation (Hard)

Find the nature of roots of quadratic equation ax² + bx + c = 0:
- Real and distinct (discriminant > 0)
- Real and equal (discriminant = 0)
- Imaginary (discriminant < 0)

Discriminant = b² - 4ac

**Hint:** Calculate discriminant = b**2 - 4*a*c. Use if-elif-else to check if > 0, == 0, or < 0.

In [ ]:
# Your code here


### Problem 8.23: Bonus Calculation (Medium)

Calculate employee bonus:
- Salary < 10000: 10% bonus
- Salary 10000-20000: 15% bonus
- Salary > 20000: 20% bonus
- If years of service > 5: Add extra 5%

**Hint:** Determine bonus percentage based on salary. If years > 5, add 5 to bonus percentage. Calculate: bonus = salary * bonus_percentage / 100.

In [ ]:
# Your code here


### Problem 8.24: Water State (Easy)

Determine water state based on temperature:
- Below 0°C: Ice
- 0-100°C: Water
- Above 100°C: Steam

**Hint:** If temp < 0: Ice, elif temp <= 100: Water, else: Steam.

In [ ]:
# Your code here


### Problem 8.25: Alphabet or Digit (Easy)

Check if a character is an alphabet or a digit.

**Hint:** Use ch.isalpha() to check alphabet, ch.isdigit() for digit.

In [ ]:
# Your code here


### Problem 8.26: Tax Calculator (Medium)

Calculate income tax:
- Income 0-250000: No tax
- Income 250001-500000: 5% tax
- Income 500001-1000000: 10% tax
- Income above 1000000: 20% tax

**Hint:** Use if-elif-else to determine tax rate. Calculate: tax = income * tax_rate / 100.

In [ ]:
# Your code here


### Problem 8.27: Number of Days Calculator (Medium)

Calculate the difference in days between two dates (simplified: just day numbers in same month).

**Hint:** Simply calculate abs(day2 - day1) for absolute difference.

In [ ]:
# Your code here


### Problem 8.28: Grade with Plus/Minus (Hard)

Assign grades with +/- modifiers:
- 97-100: A+
- 93-96: A
- 90-92: A-
- 87-89: B+
- (continue pattern for B, C, D, F)

**Hint:** Use if-elif-else with specific ranges. Start from highest (A+) and work down to F.

In [ ]:
# Your code here


### Problem 8.29: Shipping Cost Calculator (Medium)

Calculate shipping cost based on weight and distance:
- Weight < 1kg: $5
- Weight 1-5kg: $10
- Weight > 5kg: $15
- If distance > 100km: Add $5

**Hint:** Determine base cost by weight. Then check if distance > 100 and add extra cost.

In [ ]:
# Your code here


### Problem 8.30: Blood Donation Eligibility (Medium)

Check eligibility for blood donation:
- Age: 18-65
- Weight: >= 50kg
- Hemoglobin: >= 12.5 g/dL

**Hint:** Check if (18 <= age <= 65) and (weight >= 50) and (hemoglobin >= 12.5) all are True.

In [ ]:
# Your code here


---

## Section 9: Additional For Loop Practice (30 Questions)

### Problem 9.1: Print Numbers 1 to N (Easy)

Print all numbers from 1 to n.

**Hint:** Use for i in range(1, n+1): print(i)

In [ ]:
# Your code here


### Problem 9.2: Print Numbers in Reverse (Easy)

Print numbers from n to 1 in descending order.

**Hint:** Use for i in range(n, 0, -1): print(i)

In [ ]:
# Your code here


### Problem 9.3: Sum of First N Natural Numbers (Easy)

Calculate the sum: 1 + 2 + 3 + ... + n

**Hint:** Initialize sum=0. Loop from 1 to n+1, add each i to sum.

In [ ]:
# Your code here


### Problem 9.4: Product of Digits (Medium)

Calculate the product of all digits in a number.

**Hint:** Convert to string, loop through each digit: product *= int(digit). Or extract digits using modulo.

In [ ]:
# Your code here


### Problem 9.5: Print All Odd Numbers (Easy)

Print all odd numbers from 1 to n.

**Hint:** Use range(1, n+1, 2) to get only odd numbers, or check if i%2 != 0.

In [ ]:
# Your code here


### Problem 9.6: Print All Even Numbers (Easy)

Print all even numbers from 1 to n.

**Hint:** Use range(2, n+1, 2) to get only even numbers, or check if i%2 == 0.

In [ ]:
# Your code here


### Problem 9.7: Count Vowels in String (Medium)

Count the number of vowels in a given string.

**Hint:** Loop through each character. Check if char.lower() in 'aeiou', increment count.

In [ ]:
# Your code here


### Problem 9.8: Power Calculator (Easy)

Calculate x raised to power n using a for loop (don't use ** operator).

**Hint:** Initialize result=1. Loop n times: result *= x.

In [ ]:
# Your code here


### Problem 9.9: Sum of Squares (Medium)

Calculate: 1² + 2² + 3² + ... + n²

**Hint:** Loop from 1 to n+1: sum += i**2 or i*i.

In [ ]:
# Your code here


### Problem 9.10: Sum of Cubes (Medium)

Calculate: 1³ + 2³ + 3³ + ... + n³

**Hint:** Loop from 1 to n+1: sum += i**3 or i*i*i.

In [ ]:
# Your code here


### Problem 9.11: Print Multiples of Number (Easy)

Print first 10 multiples of a given number.

**Hint:** Loop from 1 to 11: print(number * i).

In [ ]:
# Your code here


### Problem 9.12: Count Digits (Easy)

Count total number of digits in a number using for loop.

**Hint:** Convert to string: for digit in str(n): count += 1. Or use len(str(n)).

In [ ]:
# Your code here


### Problem 9.13: Find Factors (Medium)

Print all factors of a given number.

**Hint:** Loop from 1 to n+1. If n % i == 0, print i (it's a factor).

In [ ]:
# Your code here


### Problem 9.14: Count Prime Numbers in Range (Hard)

Count how many prime numbers exist between 1 and n.

**Hint:** Use nested loops. Outer loop for each number, inner loop to check if prime. Count each prime found.

In [ ]:
# Your code here


### Problem 9.15: ASCII Values (Easy)

Print characters from 'A' to 'Z' with their ASCII values.

**Hint:** Loop from 65 to 91 (ASCII range). Use chr(i) to get character, i is ASCII value.

In [ ]:
# Your code here


### Problem 9.16: Harmonic Series (Medium)

Calculate sum of harmonic series: 1 + 1/2 + 1/3 + ... + 1/n

**Hint:** Loop from 1 to n+1: sum += 1/i. Use floating point division.

In [ ]:
# Your code here


### Problem 9.17: Reverse String (Easy)

Reverse a string using for loop.

**Hint:** Loop backwards through string: for i in range(len(s)-1, -1, -1): reversed_str += s[i].

In [ ]:
# Your code here


### Problem 9.18: Sum of Odd Numbers (Medium)

Calculate sum of all odd numbers from 1 to n.

**Hint:** Use range(1, n+1, 2) for odd numbers. Add each to sum.

In [ ]:
# Your code here


### Problem 9.19: Frequency Counter (Medium)

Count frequency of a specific digit in a number.

**Hint:** Convert number to string. Loop through each character, count if it matches target digit.

In [ ]:
# Your code here


### Problem 9.20: Alternating Sum (Medium)

Calculate: 1 - 2 + 3 - 4 + 5 - 6 + ... ± n

**Hint:** Loop from 1 to n+1. If i is odd: add i, if even: subtract i. Or use: sum += i if i%2==1 else -i.

In [ ]:
# Your code here


### Problem 9.21: Print Prime Factors (Hard)

Print all prime factors of a number.

**Hint:** Loop from 2 to n. For each i, while n % i == 0: print i, divide n by i. Continue with next i.

In [ ]:
# Your code here


### Problem 9.22: Check Palindrome String (Medium)

Check if a string is a palindrome using for loop.

**Hint:** Reverse the string (Problem 9.17), then compare original with reversed.

In [ ]:
# Your code here


### Problem 9.23: Binary to Decimal (Hard)

Convert a binary number (as string) to decimal using for loop.

**Hint:** Loop through binary string. For each bit: decimal = decimal * 2 + int(bit).

In [ ]:
# Your code here


### Problem 9.24: Decimal to Binary (Hard)

Convert a decimal number to binary using for loop.

**Hint:** While n > 0: append (n % 2) to result, then n = n // 2. Reverse the result.

In [ ]:
# Your code here


### Problem 9.25: Average of Numbers (Easy)

Calculate average of n numbers entered by user.

**Hint:** Loop n times to input numbers, add to sum. Average = sum / n.

In [ ]:
# Your code here


### Problem 9.26: Find Maximum (Easy)

Find the maximum among n numbers entered by user.

**Hint:** Initialize max with first number. Loop through remaining, update max if current number is larger.

In [ ]:
# Your code here


### Problem 9.27: Find Minimum (Easy)

Find the minimum among n numbers entered by user.

**Hint:** Similar to Problem 9.26, but update min if current number is smaller.

In [ ]:
# Your code here


### Problem 9.28: Compound Interest (Medium)

Calculate compound interest for n years: A = P(1 + r/100)^t

**Hint:** Use for loop to multiply principal by (1 + r/100) t times. Or use: A = P * (1 + r/100)**t.

In [ ]:
# Your code here


### Problem 9.29: Perfect Numbers in Range (Hard)

Print all perfect numbers between 1 and n.

**Hint:** Outer loop for each number. For each, use inner loop to find sum of divisors (Problem 6.3). If sum equals number, print it.

In [ ]:
# Your code here


### Problem 9.30: Nested Factorial Sum (Hard)

Calculate: 1! + 2! + 3! + ... + n!

**Hint:** Outer loop from 1 to n+1. For each i, calculate factorial using inner loop (Problem 2.2), add to sum.

In [ ]:
# Your code here


---

## Section 10: Additional While Loop Practice (30 Questions)

### Problem 10.1: Print 1 to N Using While (Easy)

Print numbers from 1 to n using while loop.

**Hint:** Initialize i=1. While i <= n: print(i), i += 1.

In [ ]:
# Your code here


### Problem 10.2: Countdown Timer (Easy)

Create a countdown from n to 1 using while loop.

**Hint:** While n > 0: print(n), n -= 1.

In [ ]:
# Your code here


### Problem 10.3: Find First Digit (Easy)

Find the first digit of a number using while loop.

**Hint:** While n >= 10: n = n // 10. When loop ends, n is the first digit.

In [ ]:
# Your code here


### Problem 10.4: Find Last Digit (Easy)

Find the last digit of a number.

**Hint:** Last digit = n % 10.

In [ ]:
# Your code here


### Problem 10.5: Remove Last Digit (Easy)

Remove the last digit from a number repeatedly until 0.

**Hint:** While n > 0: print(n), n = n // 10.

In [ ]:
# Your code here


### Problem 10.6: Sum Until Limit (Medium)

Keep adding numbers until sum exceeds a given limit.

**Hint:** i=1, sum=0. While sum <= limit: sum += i, i += 1. Print count or sum.

In [ ]:
# Your code here


### Problem 10.7: Fibonacci Until N (Medium)

Print Fibonacci numbers until they exceed n.

**Hint:** a=0, b=1. While a <= n: print(a), calculate next (a, b = b, a+b).

In [ ]:
# Your code here


### Problem 10.8: Digit Separator (Medium)

Print each digit of a number on a separate line.

**Hint:** While n > 0: print(n % 10), n = n // 10. This prints in reverse; store in list to print in order.

In [ ]:
# Your code here


### Problem 10.9: Factorial Using While (Medium)

Calculate factorial using while loop.

**Hint:** result=1, i=1. While i <= n: result *= i, i += 1.

In [ ]:
# Your code here


### Problem 10.10: Power Using While (Medium)

Calculate x^n using while loop.

**Hint:** result=1, count=0. While count < n: result *= x, count += 1.

In [ ]:
# Your code here


### Problem 10.11: Menu-Driven Program (Medium)

Create a menu that keeps showing options until user selects exit.

**Hint:** while True: display menu, get choice. If exit chosen, break. Use if-elif for options.

In [ ]:
# Your code here


### Problem 10.12: Validate Input (Medium)

Keep asking for input until user enters a valid positive number.

**Hint:** while True: get input. If n > 0, break. Otherwise continue loop.

In [ ]:
# Your code here


### Problem 10.13: Password Attempts (Medium)

Allow 3 attempts to enter correct password.

**Hint:** attempts=0. While attempts < 3: get password. If correct, break. attempts += 1.

In [ ]:
# Your code here


### Problem 10.14: GCD Using While (Medium)

Calculate GCD of two numbers using while loop.

**Hint:** Same as Problem 6.2. While b != 0: temp=b, b=a%b, a=temp.

In [ ]:
# Your code here


### Problem 10.15: Remove Specific Digit (Hard)

Remove all occurrences of a specific digit from a number.

**Hint:** Extract each digit. If digit != target, build new number: result = result * 10 + digit.

In [ ]:
# Your code here


### Problem 10.16: Keep Doubling (Easy)

Start with 1 and keep doubling until number exceeds n.

**Hint:** num=1. While num <= n: print(num), num *= 2.

In [ ]:
# Your code here


### Problem 10.17: Binary Conversion (Medium)

Convert decimal to binary using while loop.

**Hint:** Same as Problem 9.24. While n > 0: binary = str(n%2) + binary, n = n//2.

In [ ]:
# Your code here


### Problem 10.18: Octal Conversion (Medium)

Convert decimal to octal using while loop.

**Hint:** Similar to binary. While n > 0: octal = str(n%8) + octal, n = n//8.

In [ ]:
# Your code here


### Problem 10.19: Hexadecimal Conversion (Hard)

Convert decimal to hexadecimal using while loop.

**Hint:** hex_digits = '0123456789ABCDEF'. While n > 0: hex_str = hex_digits[n%16] + hex_str, n = n//16.

In [ ]:
# Your code here


### Problem 10.20: Check Automorphic Number (Hard)

An automorphic number is a number whose square ends in the number itself.
Examples: 5 (25), 6 (36), 76 (5776)

**Hint:** Calculate square. Extract last digits of square (using %) equal to number of digits in n. Compare with n.

In [ ]:
# Your code here


### Problem 10.21: Neon Number (Medium)

A neon number is one where sum of digits of its square equals the number.
Example: 9 (9² = 81, 8+1 = 9)

**Hint:** Calculate square. Find sum of digits of square (Problem 3.2). Compare with original number.

In [ ]:
# Your code here


### Problem 10.22: Spy Number (Medium)

A spy number is one where sum of digits equals product of digits.
Example: 123 (1+2+3 = 6, 1×2×3 = 6)

**Hint:** Calculate both sum and product of digits using while loop. Compare if equal.

In [ ]:
# Your code here


### Problem 10.23: Harshad Number (Medium)

A Harshad number is divisible by the sum of its digits.
Example: 18 (1+8=9, 18÷9=2)

**Hint:** Calculate sum of digits. Check if original number % sum == 0.

In [ ]:
# Your code here


### Problem 10.24: Abundant Number (Hard)

An abundant number is one where sum of proper divisors exceeds the number.
Example: 12 (1+2+3+4+6 = 16 > 12)

**Hint:** Find sum of divisors (1 to n-1). Check if sum > n.

In [ ]:
# Your code here


### Problem 10.25: Deficient Number (Hard)

A deficient number is one where sum of proper divisors is less than the number.
Example: 8 (1+2+4 = 7 < 8)

**Hint:** Find sum of divisors (1 to n-1). Check if sum < n.

In [ ]:
# Your code here


### Problem 10.26: Keep Dividing by 2 (Easy)

Keep dividing a number by 2 until it becomes odd.

**Hint:** While n % 2 == 0: n = n // 2. Print n when loop ends.

In [ ]:
# Your code here


### Problem 10.27: Amicable Numbers (Hard)

Check if two numbers are amicable (sum of proper divisors of each equals the other).
Example: 220 and 284

**Hint:** Find sum of divisors of a (call it sum_a), sum of divisors of b (sum_b). Check if sum_a == b and sum_b == a.

In [ ]:
# Your code here


### Problem 10.28: Magic Number (Hard)

A magic number is one where repeatedly summing digits eventually gives 1.
Example: 19 → 1+9=10 → 1+0=1

**Hint:** While n > 9: calculate sum of digits, n = sum. After loop, check if n == 1.

In [ ]:
# Your code here


### Problem 10.29: ISBN Validator (Hard)

Validate a 10-digit ISBN number using the check digit algorithm.

**Hint:** ISBN-10: sum = (d1*10 + d2*9 + d3*8 + ... + d10*1). Valid if sum % 11 == 0. Loop with position weight.

In [ ]:
# Your code here


### Problem 10.30: Kaprekar Number (Hard)

A Kaprekar number is one where square can be split into parts that sum to the original number.
Example: 45 (45² = 2025, 20+25 = 45)

**Hint:** Calculate square. Split into right part (last n digits) and left part (remaining). Check if left + right == original.

In [ ]:
# Your code here


---

## Section 11: Mixed Advanced Problems (10 Questions)

### Problem 11.1: Floyd's Triangle (Medium)

Print Floyd's triangle with n rows:
```
1
2 3
4 5 6
7 8 9 10
```

**Hint:** Use counter=1. Nested loops: for i (rows) and j (columns from 1 to i+1). Print counter, increment it.

In [ ]:
# Your code here


### Problem 11.2: Hollow Square Pattern (Medium)

Print a hollow square pattern:
```
* * * * *
*       *
*       *
*       *
* * * * *
```

**Hint:** For first and last row: print all stars. For middle rows: print star, spaces, star. Check if i==0 or i==n-1 or j==0 or j==n-1.

In [ ]:
# Your code here


### Problem 11.3: Spiral Matrix (Hard)

Print numbers in a spiral pattern for n×n matrix.

**Hint:** Use four boundaries (top, bottom, left, right). Fill in order: left to right (top row), top to bottom (right col), right to left (bottom row), bottom to top (left col). Shrink boundaries.

In [ ]:
# Your code here


### Problem 11.4: Tic-Tac-Toe Game (Hard)

Create a simple two-player Tic-Tac-Toe game.

**Hint:** Use 3x3 list. Loop for turns. Display board, get position, check validity, place X or O. Check win conditions (rows, columns, diagonals) after each move.

In [ ]:
# Your code here


### Problem 11.5: Twin Primes (Medium)

Find all twin prime pairs between 1 and n. Twin primes differ by 2.
Example: (3,5), (5,7), (11,13)

**Hint:** Loop through numbers. For each, check if both i and i+2 are prime. If both prime, print pair (i, i+2).

In [ ]:
# Your code here


### Problem 11.6: Matrix Addition (Medium)

Add two matrices (2D lists) entered by user.

**Hint:** Use nested loops. result[i][j] = matrix1[i][j] + matrix2[i][j] for each position.

In [ ]:
# Your code here


### Problem 11.7: Matrix Transpose (Medium)

Find the transpose of a matrix.

**Hint:** Transpose means swap rows and columns. transpose[j][i] = matrix[i][j]. Use nested loops.

In [ ]:
# Your code here


### Problem 11.8: Sieve of Eratosthenes (Hard)

Implement the Sieve of Eratosthenes algorithm to find all primes up to n.

**Hint:** Create boolean list (all True). For each number i from 2 to √n, mark all multiples of i as False (not prime). Print numbers that remain True.

In [ ]:
# Your code here


### Problem 11.9: Anagram Checker (Medium)

Check if two strings are anagrams of each other.

**Hint:** Sort both strings and compare, or count frequency of each character in both strings and compare.

In [ ]:
# Your code here


### Problem 11.10: Sudoku Validator (Hard)

Validate if a given 9×9 Sudoku solution is correct.

**Hint:** Check three conditions: (1) Each row has unique 1-9, (2) Each column has unique 1-9, (3) Each 3x3 box has unique 1-9. Use sets to check uniqueness.

In [ ]:
# Your code here


---

## Submission Checklist

Before submitting, make sure:
- [ ] All problems are attempted
- [ ] Code is properly commented
- [ ] All code cells run without errors
- [ ] Outputs are displayed correctly
- [ ] Edge cases are handled
- [ ] Variable names are meaningful
- [ ] Code follows proper indentation

---

**Good luck with your assignments!**